File: app/src/main/java/org/wikipedia/analytics/SuggestedEditsFeedFunnel.java
package org.wikipedia.analytics;

import androidx.annotation.NonNull;

import org.json.JSONObject;
import org.wikipedia.Constants.InvokeSource;
import org.wikipedia.WikipediaApp;
import org.wikipedia.descriptions.DescriptionEditActivity;

public class SuggestedEditsFeedFunnel extends Funnel {
    private static final String SCHEMA_NAME = "MobileWikiAppSuggestedEditsFeed";
    private static final int REVISION = 20437611;
    private InvokeSource source;
    private DescriptionEditActivity.Action type;

    public SuggestedEditsFeedFunnel(DescriptionEditActivity.Action type, InvokeSource source) {
        super(WikipediaApp.getInstance(), SCHEMA_NAME, REVISION, Funnel.SAMPLE_LOG_ALL);
        this.type = type;
        this.source = source;
    }

    public void start() {
        log(
                "action", "start"
        );
    }

    public void stop() {
        log(
                "action", "stop"
        );
    }

    public void editSuccess() {
        log(
                "action", "editSuccess"
        );
    }

    @Override
    protected void preprocessSessionToken(@NonNull JSONObject eventData) {
    }

    @Override
    protected JSONObject preprocessData(@NonNull JSONObject eventData) {
        preprocessData(eventData, "source", source.getName());
        preprocessData(eventData, "type", type == DescriptionEditActivity.Action.ADD_IMAGE_TAGS ? "tags"
                : type == DescriptionEditActivity.Action.ADD_CAPTION || type == DescriptionEditActivity.Action.TRANSLATE_CAPTION ? "captions" : "descriptions");
        return super.preprocessData(eventData);
    }
}

File: app/src/main/java/org/wikipedia/analytics/SuggestedEditsFeedFunnel.kt
package org.wikipedia.analytics

import org.json.JSONObject
import org.wikipedia.Constants.InvokeSource
import org.wikipedia.WikipediaApp
import org.wikipedia.descriptions.DescriptionEditActivity

class SuggestedEditsFeedFunnel(private val type: DescriptionEditActivity.Action, private val source: InvokeSource) :
        Funnel(WikipediaApp.getInstance(), SCHEMA_NAME, REVISION, SAMPLE_LOG_ALL) {

    fun start() {
        log("action", "start")
    }

    fun stop() {
        log("action", "stop")
    }

    fun editSuccess() {
        log("action", "editSuccess")
    }

    override fun preprocessSessionToken(eventData: JSONObject) {}

    override fun preprocessData(eventData: JSONObject): JSONObject {
        preprocessData(eventData, "source", source.value)
        preprocessData(eventData, "type", if (type === DescriptionEditActivity.Action.ADD_IMAGE_TAGS) "tags"
        else if (type === DescriptionEditActivity.Action.ADD_CAPTION || type === DescriptionEditActivity.Action.TRANSLATE_CAPTION) "captions"
        else "descriptions")
        return super.preprocessData(eventData)
    }

    companion object {
        private const val SCHEMA_NAME = "MobileWikiAppSuggestedEditsFeed"
        private const val REVISION = 20437611
    }
}
